// fontkit/src/TrueTypeCollection.js
// Generated by CoffeeScript 1.8.0
(function() {
  var Directory, TTFFont, TrueTypeCollection, r, tables;

  r = require('restructure');

  TTFFont = require('./TTFFont');

  Directory = require('./tables/directory');

  tables = require('./tables');

  TrueTypeCollection = (function() {
    var TTCHeader, get;

    TTCHeader = new r.VersionedStruct(r.uint32, {
      0x00010000: {
        numFonts: r.uint32,
        offsets: new r.Array(r.uint32, 'numFonts')
      },
      0x00020000: {
        numFonts: r.uint32,
        offsets: new r.Array(r.uint32, 'numFonts'),
        dsigTag: r.uint32,
        dsigLength: r.uint32,
        dsigOffset: r.uint32
      }
    });

    function TrueTypeCollection(stream) {
      this.stream = stream;
      if (this.stream.readString(4) !== 'ttcf') {
        throw new Error('Not a TrueType collection');
      }
      this.header = TTCHeader.decode(this.stream);
    }

    TrueTypeCollection.prototype.getFont = function(name) {
      var font, offset, stream, _i, _len, _ref;
      _ref = this.header.offsets;
      for (_i = 0, _len = _ref.length; _i < _len; _i++) {
        offset = _ref[_i];
        stream = new r.DecodeStream(this.stream.buffer);
        stream.pos = offset;
        font = new TTFFont(stream);
        if (font.postscriptName === name) {
          return font;
        }
      }
      return null;
    };

    get = function(key, fn) {
      return Object.defineProperty(TrueTypeCollection.prototype, key, {
        get: fn,
        enumerable: true
      });
    };

    get('fonts', function() {
      var fonts, offset, stream, _i, _len, _ref;
      fonts = [];
      _ref = this.header.offsets;
      for (_i = 0, _len = _ref.length; _i < _len; _i++) {
        offset = _ref[_i];
        stream = new r.DecodeStream(this.stream.buffer);
        stream.pos = offset;
        fonts.push(new TTFFont(stream));
      }
      return fonts;
    });

    return TrueTypeCollection;

  })();

  module.exports = TrueTypeCollection;

}).call(this);
